code/datums/components/echolocation.dm

```diff
@@ -0,0 +1,204 @@
+	if(!istype(echolocator))
+		return COMPONENT_INCOMPATIBLE
+	if(!danger_turfs)
+		danger_turfs = typecacheof(list(/turf/open/chasm, /turf/open/lava, /turf/open/floor/fakepit))
+	if(!allowed_paths)
+		allowed_paths = typecacheof(list(/turf/closed, /obj, /mob/living)) + danger_turfs - typecacheof(/obj/effect/decal)
+	if(!isnull(echo_range))
+		src.echo_range = default_echo_range
+	if(!isnull(cooldown_time))
+		src.cooldown_time = cooldown_time
+	if(!isnull(image_expiry_time))
+		src.image_expiry_time = image_expiry_time
+	if(!isnull(fade_in_time))
+		src.fade_in_time = fade_in_time
+	if(!isnull(fade_out_time))
+		src.fade_out_time = fade_out_time
+	if(!isnull(images_are_static))
+		src.images_are_static = images_are_static
+	if(ispath(color_path))
+		client_color = echolocator.add_client_colour(color_path)
+	else
+		client_color = echolocator.add_client_colour(/datum/client_colour/echolocate)
+
+	src.echo_group = echo_group || REF(src)
+	echolocator.add_traits(list(TRAIT_ECHOLOCATION_RECEIVER, TRAIT_NIGHT_VISION), echo_group) //so they see all the tiles they echolocated, even if they are in the dark
+	echolocator.become_blind()
+	echolocator.overlay_fullscreen("echo", /atom/movable/screen/fullscreen/echo)
+	START_PROCESSING(SSfastprocess, src)
+
+/datum/component/echolocation/Destroy(force, silent)
+	STOP_PROCESSING(SSfastprocess, src)
+	var/mob/living/echolocator = parent
+	echolocator.remove_traits(list(TRAIT_ECHOLOCATION_RECEIVER, TRAIT_NIGHT_VISION), echo_group)
+	echolocator.cure_blind()
+	echolocator.clear_fullscreen("echo")
+	for(var/timeframe in images)
+		delete_images(timeframe)
+	return ..()
+
+/datum/component/echolocation/process()
+	var/mob/living/echolocator = parent
+	if(echolocator.stat == DEAD)
+		return
+	echolocate()
+
+/datum/component/echolocation/proc/echolocate()
+	if(!COOLDOWN_FINISHED(src, cooldown_last))
+		return
+	COOLDOWN_START(src, cooldown_last, cooldown_time)
+	var/mob/living/echolocator = parent
+	echo_range = echo_sound_environment(echolocator, default_echo_range)
+	var/list/filtered = list()
+	var/list/seen = dview(echo_range, get_turf(echolocator.client?.eye || echolocator), invis_flags = echolocator.see_invisible)
+	for(var/atom/seen_atom as anything in seen)
+		if(!seen_atom.alpha)
+			continue
+		if(allowed_paths[seen_atom.type])
+			filtered += seen_atom
+	if(!length(filtered))
+		return
+	var/current_time = "[world.time]"
+	images[current_time] = list()
+	receivers[current_time] = list()
+	for(var/mob/living/viewer in filtered)
+		if(HAS_TRAIT_FROM(viewer, TRAIT_ECHOLOCATION_RECEIVER, echo_group))
+			receivers[current_time] += viewer
+	for(var/atom/filtered_atom as anything in filtered)
+		show_image(saved_appearances["[filtered_atom.icon]-[filtered_atom.icon_state]"] || generate_appearance(filtered_atom), filtered_atom, current_time)
+	addtimer(CALLBACK(src, PROC_REF(fade_images), current_time), image_expiry_time)
+
+/datum/component/echolocation/proc/echo_sound_environment(mob/living/H, range)
```

Comment:
Remove the single letter vars here please

Reply:
average tgcoder